openmls/src/group/mls_group/process.rs

```diff
@@ -0,0 +1,175 @@
+use std::ops::DerefMut;
+
+use mls_group::{proposals::StagedProposal, staged_commit::StagedCommit};
+use tls_codec::Serialize;
+
+use super::{proposals::ProposalStore, *};
+
+impl MlsGroup {
+    /// This function is used to parse messages from the DS.
+    /// It checks for syntactic errors and makes some semantic checks as well.
+    /// If the input is a [MlsCiphertext] message, it will be decrypted.
+    /// Returns an [UnverifiedMessage] that can be inspected and later processed in
+    /// [self::process_unverified_message()].
+    /// Checks the following semantic validation:
+    ///  - ValSem2
+    ///  - ValSem3
+    ///  - ValSem4
+    ///  - ValSem5
+    ///  - ValSem6
+    ///  - ValSem7
+    ///  - ValSem9
+    pub fn parse_message(
+        &mut self,
+        message: MlsMessageIn,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<UnverifiedMessage, MlsGroupError> {
+        // Checks the following semantic validation:
+        //  - ValSem2
+        //  - ValSem3
+        self.validate_framing(&message)?;
+
+        // Checks the following semantic validation:
+        //  - ValSem6
+        let decrypted_message = match message.wire_format() {
+            WireFormat::MlsPlaintext => DecryptedMessage::from_inbound_plaintext(message)?,
+            WireFormat::MlsCiphertext => DecryptedMessage::from_inbound_ciphertext(
+                message,
+                self.ciphersuite(),
+                backend,
+                self.epoch_secrets(),
+                self.secret_tree_mut().deref_mut(),
+            )?,
+        };
+
+        let mut credential = None;
+
+        // Checks the following semantic validation:
+        //  - ValSem4
+        //  - ValSem5
+        //  - ValSem7
+        //  - ValSem9
+        self.validate_plaintext(decrypted_message.plaintext())?;
+
+        // Extract the credential if the sender is a member
+        let sender = decrypted_message.sender();
+        if sender.is_member() {
+            let sender_index = sender.to_leaf_index();
+
+            // Unwrapping here is safe, because we know the leaf node exists and is not blank
+            credential = Some(
+                self.tree().nodes[sender_index]
+                    .key_package
+                    .as_ref()
+                    .unwrap()
+                    .credential()
+                    .clone(),
+            );
+        }
+
+        Ok(UnverifiedMessage::from_decrypted_message(
+            decrypted_message,
+            credential,
+        ))
+    }
+
+    /// This processing function does most of the semantic verifications.
+    /// It returns a [ProcessedMessage] enum.
+    /// Checks the following semantic validation:
+    ///  - ValSem8
+    ///  - ValSem10
+    ///  - ValSem100
```

Comment:
In the function body, I could only find references to ValSem8 and ValSem10. Did I miss something?

Reply:
That should be fixed now.